src/keymgmt.c

```diff
@@ -2161,6 +2166,479 @@ const OSSL_DISPATCH p11prov_ed448_keymgmt_functions[] = {
     { 0, NULL },
 };
 
+DISPATCH_KEYMGMT_FN(mldsa, new);
+DISPATCH_KEYMGMT_FN(mldsa, free);
+DISPATCH_KEYMGMT_FN(mldsa, has);
+DISPATCH_KEYMGMT_FN(mldsa, match);
+DISPATCH_KEYMGMT_FN(mldsa, import_types);
+DISPATCH_KEYMGMT_FN(mldsa, export);
+DISPATCH_KEYMGMT_FN(mldsa, export_types);
+DISPATCH_KEYMGMT_FN(mldsa, get_params);
+DISPATCH_KEYMGMT_FN(mldsa, gettable_params);
+DISPATCH_KEYMGMT_FN(mldsa_44, gen_init);
+DISPATCH_KEYMGMT_FN(mldsa_65, gen_init);
+DISPATCH_KEYMGMT_FN(mldsa_87, gen_init);
+DISPATCH_KEYMGMT_FN(mldsa, gen);
+DISPATCH_KEYMGMT_FN(mldsa, gen_settable_params);
+
+static void *p11prov_mldsa_new(void *provctx)
+{
+    P11PROV_CTX *ctx = (P11PROV_CTX *)provctx;
+    CK_RV ret;
+
+    P11PROV_debug("mldsa new");
+
+    ret = p11prov_ctx_status(ctx);
+    if (ret != CKR_OK) {
+        return NULL;
+    }
+
+    return p11prov_obj_new(provctx, CK_UNAVAILABLE_INFORMATION,
+                           CK_P11PROV_IMPORTED_HANDLE,
+                           CK_UNAVAILABLE_INFORMATION);
+}
+
+static void p11prov_mldsa_free(void *key)
+{
+    P11PROV_debug("mldsa free %p", key);
+    p11prov_obj_free((P11PROV_OBJ *)key);
+}
+
+static void *p11prov_mldsa_load(const void *reference, size_t reference_sz)
+{
+    P11PROV_debug("mldsa load %p, %ld", reference, reference_sz);
+    return p11prov_common_load(reference, reference_sz, CKK_ML_DSA);
+}
+
+static int p11prov_mldsa_has(const void *keydata, int selection)
+{
+    P11PROV_OBJ *key = (P11PROV_OBJ *)keydata;
+
+    P11PROV_debug("mldsa has %p %d", key, selection);
+
+    if (key == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    if (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) {
+        if (p11prov_obj_get_class(key) != CKO_PRIVATE_KEY) {
+            return RET_OSSL_ERR;
+        }
+    }
+
+    /* We always return OK when asked for a PUBLIC KEY, even if we only have a
+     * private key, as we can try to fetch the associated public key as needed
+     * if asked for an export (main reason to do this), or other operations */
+
+    return RET_OSSL_OK;
+}
+
+static int p11prov_mldsa_match(const void *keydata1, const void *keydata2,
+                               int selection)
+{
+    P11PROV_debug("mldsa match %p %p %d", keydata1, keydata2, selection);
+
+    return p11prov_common_match(keydata1, keydata2, CKK_ML_DSA, selection);
+}
+
+static void *p11prov_mldsa_gen_init_int(void *provctx, int selection,
+                                        const OSSL_PARAM params[],
+                                        CK_ML_DSA_PARAMETER_SET_TYPE param_set)
+{
+    struct key_generator *ctx = NULL;
+    int ret;
+
+    P11PROV_debug("ML-DSA gen_init %p", provctx);
```

Comment:
nit: half of the debug logs have ML-DSA and the other mldsa -- would be good to make it systematic so analysis of the debug log is easier when searching for what is going on with the given operation.